Test/dafny0/NoMoreAssume2Less2.dfy

```diff
@@ -0,0 +1,199 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// This file tests various ways of writing the equivalent of "assume false;" or "assert false;" inside a loop.
+// It used to be that Dafny's translation of such statements into Boogie could cause Boogie to cut off the
+// loop back edge, which would turn a loop into something that wasn't a loop. In particular, it would cause
+// the expected "havoc loop_targets;" statement never to be generated, whose dramatic impact on what gets
+// verified would be both mysterious and surprising to most Dafny users.
```

Comment:
Could you provide an example of this dramatic impact?

Reply:
If you look at a program snippet
``` dafny
var i := 0;
while i < 100
  invariant 0 <= i <= 100
{
  assert i != 18;
  assume false;
  i := i + 1;
}
```
you would expect the `assert i != 18;` to generate an error. This is because you see the loop, and you know that a verifier checks the loop body to be correct for all values for `i` that satisfy the invariant.

A dramatically different program is
``` dafny
var i := 0;
assert 0 <= i <= 100;
if i < 100 {
  assert i != 18;
  assume false;
  i := i + 1;
  assume false;
}
```
Here, `assert i != 18;` is executed at most once. If control ever reaches it, `i` is certain to have the value `0`, so you can prove the assertion to be correct.

If you remove the loop back edge from the first of these two programs (which, as I've described above, is what Boogie's early will do when they encounter the `assume false;` in the loop body), then you get the second program. So, blocking the back edge with an `assume false;` has a dramatic impact on the program that is verified.

Reply:
> you would expect the assert i != 18; to generate an error.

That seems subjective. If we replace the `assert` and `assume` with runtime assertions, then `assert i != 18` wouldn't fail because `assert false` fails first, so in that sense I wouldn't expect a verification failure for `assert i != 18`.

I'm guessing you're confident that most Dafny users expect `assert i != 18` to give an error here so that seems fine then.

Reply:
The rule is that you reason about a loop in terms of its:
* loop guard,
* loop invariant,
* frame (which includes the explicit or inherited `modifies` clause, as well as those local variables that may be modified by the loop body), and
* `decreases` clause.

From the context surrounding the loop, this reasoning technique is independent of the loop body. This is what you'd expect in verification (that is, it is the standard rule that one uses, though one could in principle invent other loop rules). The rule is similar to how a caller reasons about a method call from the specification of the method, not its body.

Reply:
Okay :)